debezium-embedded/src/main/java/io/debezium/embedded/ConnectorEngine.java

```diff
@@ -0,0 +1,1548 @@
+ * events produced by the connectors.
+ * <p>
+ * <h2>Class loaders</h2>
+ * <p>
+ * When you create a ConnectorEngine, you can optionally provide a class loader that the engine will use as a default for
+ * each of the connectors. If none is provided the engine will simply use the same class loader that loaded the ConnectorEngine
+ * class.
+ * <p>
+ * You can also provide a class loader when {@link #addConnector(Configuration, ConnectorCallback, ClassLoader) adding} each
+ * connector, and the engine will then use that class loader to load the connector's classes.
+ * <p>
+ * <h2>Callbacks</h2>
+ * <p>
+ * The engine logs very little information, so all feedback as to when the connectors and tasks start, stop, or fail are
+ * sent through <em>callbacks</em> that are provided when {@link #addConnector(Configuration, ConnectorCallback, ClassLoader)
+ * defining} connectors. The {@link ConnectorCallback} interface has methods that include the connector name, so the same
+ * callback object can be used for multiple connectors.
+ * <p>
+ * <h2>Consuming events</h2>
+ * <p>
+ * Each connector engine provides a single {@link EventStream} of all events generated by the deployed connectors.
+ * The {@link #poll(long, TimeUnit)} method is the primary means of accessing these events, and each list of events returned
+ * by this method will contain events from a single connector. The engine will provide back-pressure on the connectors
+ * should the events not be consumed as quickly as they are produced, and this back-pressure can be controlled by
+ * adjusting the {@link #MAX_QUEUE_SIZE size of the engine's queue}.
+ * <p>
+ * However, reading the events from the stream does not mean that the consuming application has completely processed or consumed
+ * that event. Instead, the consuming application is expected to consume them in the same order in which they are read
+ * and is <b>required</b> to call {@link ChangeEvent#commit()} on every event when it has been fully processed. Only when this
+ * is done will the engine record the <em>offset</em> for the event, so that upon restart previously committed events
+ * will not be seen again.
+ * <p>
+ * <h2>Offsets</h2>
+ * <p>
+ * The engine records offsets in an <em>offset store</em> defined in the engine's configuration, and two stores are available
+ * out of the box. The first is a {@link FileOffsetBackingStore} that stores all offsets for all connectors in a single local
+ * file, and the application running the engine is expected to ensure this file is available and persisted properly.
+ * The second is the {@link KafkaOffsetBackingStore} that stores all offsets for all connectors in a Kafka topic, which provides
+ * more durable storage assuming the Kafka broker is replicated sufficiently. Of course any {@link OffsetBackingStore}
+ * implementation can be used.
+ * 
+ * @author Randall Hauch
+ */
+public class ConnectorEngine implements EventStream {
```

Comment:
this could probably use the `@ThreadSafe` annotation, since the implementation seems to be that